design/src/rfcs/rfc0012_constraint_traits.md

```diff
@@ -0,0 +1,424 @@
+
+> restricts string shape values to a specified regular expression.
+
+We will implement this by using the `regex`'s crate [`is_match`]. We will use
+[`lazy_static!`] to compile the regex only the first time it is required.
+
+`uniqueItems` trait
+-------------------
+
+The [`uniqueItems` trait]
+
+> indicates that the items in a [`List`] MUST be unique.
+
+If the list shape is `sparse`, more than one `null` value violates this
+constraint.
+
+We will enforce this by copying _references_ to the `Vec`'s elements into a
+`HashSet` and checking that the sizes of both containers coincide. Smithy
+allows you to apply `uniqueItems` to _any_ list, [even if its members are not
+equipped with an equivalence relation], so we would have to manually check that
+the list's members are all [`Eq`].
+
+[`Eq`]: https://doc.rust-lang.org/std/cmp/trait.Eq.html
+[even if its members are not equipped with an equivalence relation]: https://github.com/awslabs/smithy/issues/1093
+[`List`]: https://awslabs.github.io/smithy/1.0/spec/core/model.html#list
+[`uniqueItems` trait]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#uniqueitems-trait
+[`is_match`]: https://docs.rs/regex/latest/regex/struct.Regex.html#method.is_match
+[`pattern` trait]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#pattern-trait
+[`aws_smithy_types::Blob`]: https://docs.rs/aws-smithy-types/latest/aws_smithy_types/struct.Blob.html
+[`aws_smithy_http_server::rejection::SmithyRejection`]: https://docs.rs/aws-smithy-http-server/latest/aws_smithy_http_server/rejection/enum.SmithyRejection.html
+[`aws_smithy_http_server::rejection::Deserialize`]: https://docs.rs/aws-smithy-http-server/latest/aws_smithy_http_server/rejection/struct.Deserialize.html
+[#1187]: https://github.com/awslabs/smithy-rs/issues/1187
+[`lazy_static!`](https://docs.rs/lazy_static/1.4.0/lazy_static/)
+
+Trait precedence and naming of the tuple struct
+-----------------------------------------------
+
+From [the spec]:
+
+> Some constraints can be applied to shapes as well as structure members. If a
+> constraint of the same type is applied to a structure member and the shape
+> that the member targets, the trait applied to the member takes precedence.
+
+```smithy
+structure ShoppingCart {
+    @range(min: 7, max:12)
+    numberOfItems: PositiveInteger
+}
+
+@range(min: 1)
+integer PositiveInteger
+```
+
+In the above example,
+
+> the range trait applied to numberOfItems takes precedence over the one
+> applied to PositiveInteger. The resolved minimum will be 7, and the maximum
+> 12.
+
+When the constraint trait is applied to a member shape, the tuple struct's name
+will be the PascalCased name of the member shape, `NumberOfItems`.
+
+[the spec]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html?highlight=enum#trait-precedence
+
+Unresolved questions
+--------------------
+
+1. Should we code-generate unsigned integer types (`u16`, `u32`, `u64`) when
```

Comment:
Very good point. I'm of the opinion that if you add / remove a constraint, you're backwards-incompatibly affecting your service's behavior, so a backwards-incompatible change in server user code seems appropriate.

Reply:
Yeah, I think breaking the build is fine if a constraint has been tightened, but if it's been loosened, that's really unfair. In my opinion, the types in the target language should be derived solely from the smithy types, and they shouldn't be impacted by the constraint traits outside of `@required`/`@default`.

Reply:
Would it be possible to regard those "custom Rust features" as Smithy traits that live outside of the official Smithy implementation?

For example one could decide to create the following trait
```smithy
@trait(selector: "integer")
structure rustInteger {
      @required
      int: String
}

@rustInteger("u8")
Integer MyInteger
```
and them modify smithy-rs to support that `rustInteger` trait?

That would allow a team using Rust to apply custom traits to a model shared with other team via [the external trait application](https://awslabs.github.io/smithy/1.0/spec/core/model.html#applying-traits-externally).

Reply:
@oddg Certainly, you can do all sorts of custom traits in Smithy. I think I like more the idea of having an `@unsigned` trait for this. 

---

One thing to note is that if we were to generate unsigned types we would be expanding the range of accepted values by Smithy. For example, an [`integer` shape in Smithy](https://awslabs.github.io/smithy/1.0/spec/core/model.html#simple-shapes) is a 32-bit integer with range from -2^31 to (2^31)-1. That's _exactly_ the range of [`i32`](https://doc.rust-lang.org/std/primitive.i32.html) in Rust. If we tack on `@unsigned` or `@range(min: 0)` to the shape, it's unclear that it's too helpful to generate an [`u32`](https://doc.rust-lang.org/std/primitive.u32.html), because the upper bound of that type is (2^32)-1. The wrapper tuple `MyInteger(u32)` struct would only have to check the upper bound.

I'm now inclined to drop this feature and have `range` on integers _not_ affect whether we generate signed or unsigned integers.

Reply:
if you wanted to, you could impl an `.unsigned()` method when the model allowed to enable converting from i32 to u32 fearlessly

Reply:
Worth mentioning that the clients should also support this if the server does, or there will be issues when a server responds with something outside of the signed int range.